libmscore/layout.cpp

```diff
@@ -3764,6 +3764,8 @@ void Score::layoutSystemElements(System* system, LayoutContext& lc)
                                     system->staff(f->note()->chord()->vStaffIdx())->skyline().add(r);
                                     }
                               }
+                              if (!fingerings.empty())
+                                    s->createShapes();
```

Comment:
One more performance-related suggestion: it shouldn't be necessary to completely regenerate shapes for the segment on each and every chord that contains fingering.  I suppose ideally we could simply add the fingering shape itself to the segment shape directly, but that might be easier said than done.  So maybe it really is better just set a flag here indicating that shapes need to be regenerated, but not actually do it until we're done with the loop through the elist.  Sorry, I know that kind of contradicts my previous suggestion.

Reply:
I'm not sure having the <code>createShapes</code> inside the loop will do that much harm, compared with what layout in general already is doing but, agreed, all bits can help. Or worse, all small bit can add up and become a nightmare, some point in time.
I did some tests putting <code>createShapes</code> outside the <code>elist</code> loop and I didn't see any issues popping up.
Generating fingering shapes directly at the segment seems tricky to me. Now <code>createShapes</code> just removes all shapes and builds up everything from scratch. By making an exception for fingering shapes I think the code becomes complex and error prone. When looking into this issue I already saw the methods are called several time (and from several places). It will be very difficult to make sure all shapes are always there when needed. I might require a full rework to make this path worthwhile to follow.

Reply:
For the record, createShapes goes through all staves, so in a score with multiple staves that include fingerings, that's a fair amount of rework to regenerate the shapes multiple times for each segment.  That's why it's potentially important.  If there are other places where we can make similar optimizations, great, we should do that too eventually.

Unfortunately, simply moving the code out of the loop as you did is likely to be problematic. 
 Now the fingering array is not cleared from chord to chord.  I haven't tested, but I think your code could produce bad effects if there are multiple chords with fingering on the same segment, as now the fingerings will be laid out multiple times each.

It would need to be as I said, set a flag, check the flag later so the actual processing of fingerings is not affected.  Or, perhaps, put the code back here you had it last, but add a staffIdx parameter to regenerate the shapes just for the staff in question.  Hmm, that's possibly actually the best, it prevents the need to regenerate shapes for unaffected staves at all.  Only a staff with multiple voices each including a fingering would be re-shaped unnecessarily.  Fixing that is possible too but probably more trouble than it's worth.

I agree trying to stuff the fingering shapes in manually is not advisable given the current code organization.  It's the sort of thing we might consider streamlining someday though,

Reply:
I see your point of creating a potential performance issue. Not sure what the best solution will be but I will certainly have a look (after I finished some other issues I'm working on right now).

Reply:
> Unfortunately, simply moving the code out of the loop as you did is likely to be problematic. Now the fingering array is not cleared from chord to chord. 

I don't that is an issue here. It seems the layout of the fingering doesn't take any notes into account at all. That means the code can stay outside the loop. An extra optimization could be by create the shapes for only these staffId's which contains fingerings. This implemented in the new push.

Reply:
To be clear, my concern wasn't about moving the code out, but with the fact that the array wasn't being cleared, meaning which would have resulted in multiple layouts of the same fingering.  The code for how the fingerings stack above/below the chord depends on being called only once, and that is also we need to explicitly clear the old layout information a few lines above this whole loop.  It's kind of a hack, but was the only solution I found to the issue of having fingerings being attached to individual notes but needing to be laid out relative to the chord as a whole.

Anyhow, I think what you have now is pretty much ideal.  The fingering layout itself isn't affected, we just regenerate the shapes for the affected staves.  Use of a "set" for this is clever, it solves the multiple voice problem nicely, better than my suggestion of setting and testing a flag.

Reply:
Altogether it seems to me this is the best solution for now.
With all the testing I found a shift after the first relayout after loading the score (indeed, the famous one) but this shift is **not** caused by this PR. As a matter of fact, it was already there in 3.0. But this will be subject for another PR.